File: Templatium.Docx/DocxTemplater.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;
using Templatium.Docx.Contents;
using Templatium.Docx.Extensions;
using Templatium.Docx.Processors;

namespace Templatium.Docx
{
    public class DocxTemplater : IDocxTemplater
    {
        private readonly Dictionary<Type, IProcessor> _processors = new();

        public IReadOnlyDictionary<Type, IProcessor> Processors => _processors;

        public IProcessor? GetProcessor(Type targetType)
            => _processors.TryGetValue(targetType, out var processor)
                ? processor
                : null;

        public void AddProcessor(IProcessor processor)
            => _processors.Add(processor.TargetType, processor);

        public void GenerateDocument(WordprocessingDocument input, ContentContainer contents)
        {
            var collection = GetAllSdtElements(input);

            foreach (var sdt in collection)
            {
                var tag = sdt.SdtProperties.Descendants<Tag>().FirstOrDefault();
                if (tag == null)
                    throw new Exception("Not found any tag inside std element");

                var content = contents.GetContent(tag.Val.Value);
                if (content == null)
                    continue;

                if (!_processors.TryGetValue(content.GetType(), out var processor))
                    throw new Exception($"Not found processor for {content.GetType()}");

                if (processor.CanFill(this, input, sdt, content))
                    processor.FillContent(this, input, sdt, content);
            }
        }

        public void DeleteContentControls(WordprocessingDocument doc)
        {
            var sdtElements = GetAllSdtElements(doc);

            for (var i = sdtElements.Count - 1; i >= 0; i--)
            {
                var sdt = sdtElements[i];

                var content = sdt.Descendants().FirstOrDefault(o => o.LocalName == "sdtContent");

                if (content != null)
                {
                    var prev = sdt.PreviousSibling();
                    foreach (var child in content.ChildElements)
                        prev = prev.InsertAfterSelf((OpenXmlElement) child.Clone());
                }

                sdt.Remove();
            }
        }

        private List<SdtElement> GetAllSdtElements(WordprocessingDocument doc)
        {
            var collection = doc.MainDocumentPart.Document.Body.FindByName("sdt").ToList();

            collection.AddRange(
                doc.MainDocumentPart.FooterParts
                    .SelectMany(c => c.Footer.Descendants<SdtElement>())
            );
            collection.AddRange(
                doc.MainDocumentPart.HeaderParts
                    .SelectMany(c => c.Header.Descendants<SdtElement>())
            );

            return collection.OfType<SdtElement>().ToList();
        }
    }
}

File: Templatium.Docx/DocxTemplater.fs
namespace Templatium.Docx

open DocumentFormat.OpenXml
open DocumentFormat.OpenXml.Packaging
open DocumentFormat.OpenXml.Wordprocessing
open Templatium.Docx

module DocxTemplater =
    let inline private getAllSdtNodesFromNode (node: OpenXmlElement) =
        seq {
            match node with
            | :? SdtElement as sdt -> yield sdt
            | _ -> ()

            yield! OpenXmlHelpers.findDescendantsByName<SdtElement> node Constants.sdt
        }

    let private getAllSdtNodesFromDoc (doc: WordprocessingDocument) =
        let sdts = ResizeArray()

        sdts.AddRange(getAllSdtNodesFromNode doc.MainDocumentPart.Document.Body)

        let inline findSdts (part: OpenXmlPart) =
            OpenXmlHelpers.findDescendantsByName part.RootElement Constants.sdt

        sdts.AddRange(
            doc.MainDocumentPart.HeaderParts
            |> Seq.collect findSdts
        )

        sdts.AddRange(
            doc.MainDocumentPart.FooterParts
            |> Seq.collect findSdts
        )

        if doc.MainDocumentPart.FootnotesPart <> null then
            sdts.AddRange(findSdts doc.MainDocumentPart.FootnotesPart)

        if doc.MainDocumentPart.EndnotesPart <> null then
            sdts.AddRange(findSdts doc.MainDocumentPart.EndnotesPart)

        sdts

    let fillNode
        (processors: IProcessor seq)
        (contents: IContent seq)
        (doc: WordprocessingDocument)
        (node: OpenXmlElement)
        =
        let sdts = getAllSdtNodesFromNode node

        let metadata =
            { Processors = processors
              Document = doc }

        for sdt in sdts do
            let titleNode = OpenXmlHelpers.findFirstNodeByName<SdtAlias> sdt Constants.alias

            match titleNode with
            | None -> ()
            | Some alias ->
                let contentOpt =
                    contents
                    |> Seq.tryFind (fun c -> c.Title = alias.Val.Value)

                match contentOpt with
                | None -> ()
                | Some content ->
                    processors
                    |> Seq.tryFind (fun p -> p.CanFill content sdt metadata)
                    |> Option.iter (fun p -> p.Fill content sdt metadata)

        ()


    let fillDocument (processors: IProcessor seq) (contents: IContent seq) (doc: WordprocessingDocument) =
        fillNode processors contents doc doc.MainDocumentPart.Document.Body

        let fillPart (part: OpenXmlPart) =
            fillNode processors contents doc part.RootElement

        doc.MainDocumentPart.HeaderParts
        |> Seq.iter fillPart

        doc.MainDocumentPart.FooterParts
        |> Seq.iter fillPart

        if doc.MainDocumentPart.FootnotesPart <> null then
            fillPart doc.MainDocumentPart.FootnotesPart

        if doc.MainDocumentPart.EndnotesPart <> null then
            fillPart doc.MainDocumentPart.EndnotesPart

        doc


    type private NearNode =
        | PreviousSibling of OpenXmlElement
        | NextSibling of OpenXmlElement
        | Parent of OpenXmlElement

    let deleteContentControls (doc: WordprocessingDocument) =
        let sdts = getAllSdtNodesFromDoc doc

        for i = sdts.Count - 1 downto 0 do
            let sdt = sdts[i]

            let sdtContentOpt = OpenXmlHelpers.findFirstNodeByName sdt Constants.sdtContent

            match sdtContentOpt with
            | Some sdtContent when sdtContent.ChildElements.Count > 0 ->

                // We need to find node relative to the first child when sdt node deleted
                let nearNode =
                    match sdt.PreviousSibling() with
                    | null ->
                        match sdt.NextSibling() with
                        | null -> Parent sdt.Parent
                        | next -> NextSibling next
                    | previous -> PreviousSibling previous

                let firstChild = sdtContent.ChildElements[0]

                let mutable prev =
                    match nearNode with
                    | PreviousSibling prevSib -> prevSib.InsertAfterSelf(firstChild.CloneNode(true))
                    | NextSibling nextSib -> sdt.Parent.InsertBefore(firstChild.CloneNode(true), nextSib)
                    | Parent parent -> parent.AppendChild(firstChild.CloneNode(true))

                sdtContent.ChildElements
                |> Seq.skip 1
                |> Seq.iter (fun child -> prev <- prev.InsertAfterSelf(child.CloneNode(true)))

                ()
            | _ -> ()

            sdt.Remove()

        ()
